File: web/src/lib/stores/session.ts
import { writable, derived } from 'svelte/store';
import type { SessionState } from '$types/session';

// Session store - manages user session state
function createSessionStore() {
	const { subscribe, set, update } = writable<SessionState>({
		database: null,
		company: null,
		userID: null,
		userName: null,
		userFullName: null,
		language: 'en-US',
		isAuthenticated: false,
		isLoading: false
	});

	return {
		subscribe,

		// Initialize session from API
		async initialize() {
			update(s => ({ ...s, isLoading: true }));

			try {
				const response = await fetch('/api/session');
				if (response.ok) {
					const data = await response.json();
					set({
						database: data.database,
						company: data.company,
						userID: data.user_id,
						userName: data.user_name,
						userFullName: data.user_full_name,
						language: data.language || 'en-US',
						isAuthenticated: !!data.user_id,
						isLoading: false
					});
				} else {
					update(s => ({ ...s, isLoading: false }));
				}
			} catch (error) {
				console.error('Failed to initialize session:', error);
				update(s => ({ ...s, isLoading: false }));
			}
		},

		// Set language
		setLanguage(language: string) {
			update(s => ({ ...s, language }));
		},

		// Clear session (logout)
		clear() {
			set({
				database: null,
				company: null,
				userID: null,
				userName: null,
				userFullName: null,
				language: 'en-US',
				isAuthenticated: false,
				isLoading: false
			});
		}
	};
}

export const session = createSessionStore();

// Derived stores
export const currentLanguage = derived(session, $session => $session.language);
export const isAuthenticated = derived(session, $session => $session.isAuthenticated);
export const currentCompany = derived(session, $session => $session.company);

File: backend/api/handlers/session.go
package handlers

import (
	"github.com/gofiber/fiber/v2"
	apitypes "github.com/hansjlachmann/openerp/backend/api/types"
)

// SessionHandler handles session-related API requests
type SessionHandler struct {
	// Can add dependencies here if needed
}

// NewSessionHandler creates a new session handler
func NewSessionHandler() *SessionHandler {
	return &SessionHandler{}
}

// GetSession returns the current session information
// GET /api/session
func (h *SessionHandler) GetSession(c *fiber.Ctx) error {
	c.Set("Cache-Control", "no-store")

	sess := getSession(c)
	if sess == nil {
		// No active session - return empty session
		response := apitypes.NewSuccessResponse(&apitypes.SessionResponse{
			Database:     "",
			Company:      "",
			UserID:       "",
			UserName:     "",
			UserFullName: "",
			Language:     "en-US",
		})
		return c.JSON(response)
	}

	// Return session data
	dbPath := ""
	if db := sess.GetDatabase(); db != nil {
		dbPath = db.GetDatabasePath()
	}

	// Normalize language code (Code type uppercases to "NB-NO", but we need "nb-NO" format)
	language := normalizeLanguageCode(sess.GetLanguage())
	if language == "" {
		language = "en-US"
	}

	sessionData := &apitypes.SessionResponse{
		Database:     dbPath,
		Company:      sess.GetCompany(),
		UserID:       sess.GetUserID(),
		UserName:     sess.GetUserName(),
		UserFullName: sess.GetUserName(), // Session doesn't have full name, use username
		Language:     language,
	}

	response := apitypes.NewSuccessResponse(sessionData)
	return c.JSON(response)
}
